---
output: html_document
editor_options: 
  chunk_output_type: console
---



### Installing LDpred-funct

# on my mac
```{bash}
conda create -n ldpredfunct 
conda activate ldpredfunct
pip install plinkio
pip install h5py scipy
```



### Input files

# 1. Plink files. **One plink file (binary PED format) per chromosome** from the validation (insert the character "[1:22]" instead of the chromsome numbers). These are used as LD-reference panel and later to compute PRS.
#convert files to ped:
```{bash}
conda activate ldpredfunct
plink2 --bgen ../private_input_files/BB_imputed_data_sample/c22_b0_testsubsample.bgen ref-first \
        --rm-dup force-first --make-bed \
        --out ../private_input_files/BB_imputed_data_sample/ped/c22_b0_testsubsample
```


#   - Example: plinkfile="my_plink_files/Final.chrom[1:22]"
 
# 2. File with functional enrichments (see Gazal et al 2017 Nat Genet and Finucane et al 2015 Nat Genet).
#     - First you will need to estimate the per-SNP heritability inferred using S-LDSC (see instructions [here](https://github.com/bulik/ldsc/wiki/Partitioned-Heritability)) under the baselineLD model (you can download Baseline-LD annotations [here](https://data.broadinstitute.org/alkesgroup/LDSCORE/)), and use the summary statistics that will be later used as training. When running S-LDSC make sure to use the --print-coefficients flag to get the regression coefficients.

## 3- The following command will allow you to partition heritability:

# on mamut
```{bash}
storage 
cd ldsc_py2.7
conda activate ldsc

python /mnt/storage/emanuele/ldsc_py2.7/ldsc.py \
	--h2 gwas/formatted_hapmap/PGC3_SCZ_wave3_european_autosome_v3_py3.sumstats.gz\
	--w-ld-chr data/weights_Osimo/weights.\
	--overlap-annot \
	--frqfile-chr data/1000G_EUR_Phase3_plink/1000G.EUR.QC.\
	--ref-ld-chr data/baseline_Osimo/baseline.\
	--print-coefficients\
	--out LDpredfunct_out/2023_06_01_baseline
```


    
#     - After running S-LDSC:
#       1. Get h2g estimate from the *.log file.
Total Observed scale h2: 0.3661 (0.0145)

# on mac Rstudio:

#       2. Get the regression coefficients from the *.results file (column 8). Divide the regression coeffients by h2g, define it as 
#           T=tau/h2g which is a vector of dimension Cx1, where C is the total number of annotations.
```{r}
library(tidyverse)

coeffs = data.table::fread("results/LDpredfunct_out/2023_06_01_baseline.results", select = c("Coefficient","Category"))%>% 
  dplyr::filter(!grepl("Transcribed_Hoffman",Category)) %>% dplyr::select(Coefficient)
head(coeffs)

T = coeffs/0.3661
head(T)

#obtain baseline LDPred categories names
(names = data.table::fread("results/LDpredfunct_out/2023_06_01_baseline.results", select = "Category") %>% 
    mutate(Category = gsub(x = Category, pattern = "L2_0", replacement = "") ) %>% 
    dplyr::filter(!grepl("Transcribed_Hoffman",Category))
)
#       3. From the baselineLD annotations downloaded from [here](https://data.broadinstitute.org/alkesgroup/LDSCORE/), read the annotations file 
#         baselineLD.*.annot.gz, and only keep the annotations columns (i.e. remove first 4 columns). Call this matrix X, with dimensions MxC, where M is the 
#         number of SNPs and C is the total number of annotations.
#create a list of the files from your target directory
(file_list <- list.files(path="baselineLD_v2.1_annots/", pattern = "baselineLD.*.annot.gz", full.names = TRUE))
#initiate a blank data frame, each iteration of the loop will append the data from the given file to this variable
X <- data.frame()
SNPs <- data.frame()
for (i in 1:length(file_list)){
  print (i)
  temp_data <- data.table::fread(file_list[i], #"baselineLD_v2.1_annots//baselineLD.1.annot.gz",
                                 select = c(names$Category))
  temp_SNPs<- data.table::fread(file_list[i], #"baselineLD_v2.1_annots//baselineLD.1.annot.gz",
                                select = "SNP")
  X <- rbind(X, temp_data) 
  SNPs <- rbind(SNPs, temp_SNPs) #for each iteration, bind the new data to the building X
  rm(temp_data, temp_SNPs)
}
head(cbind(SNPs, X))

#       4. Define the expected per-SNP heritability as a Mx1 vector (sigma2_i from the LDpred-funct manuscript) as the result from multiplying 
      # the matrix X times T.
sigma = as.matrix(X) %*% T$Coefficient
head(sigma)
#linear algebra chapters

functfile <- cbind(SNPs,sigma=sigma)
head(functfile)
```

#     - Format of FUNCTFILE:
#     - Column 1: SNP ID
#     - Column 2: per-SNP heritability

#   - Use flag: --FUNCT_FILE

# 3. Summary statistics file. Please check that the summary statistics contains a column for each of the following field (header is important here, 
#     important fields are highlighted in bold font, the order of the columns it is not important).
#     - **CHR**   Chromosome
#     - **SNP**   SNP ID
#     - **BP**    Physical position (base-pair)
#     - **A1**    Minor allele name (based on whole sample) 
#     - **A2**    Major allele name 
#     - **P**     Asymptotic p-value  
#     - **BETA**  Effect size
#     - **Z**     Z-score (default). If instead of Z-score the Chi-square statistic is provided, use the flag --chisq, and **CHISQ** as column field.
```{r}
d = data.table::fread("../private_input_files/GWAS/PGC3_SCZ_wave3.european.autosome.public.v3.tsv.gz")
names(d)
d %>% select(CHR=CHROM, SNP=ID, BP=POS, A1, A2, P=PVAL, )
```
    

#   - Use flag: --ssf

# 4. Phenotype file.
#   - Format:
#     - Column 1: FID
#     - Column 2: phenotype
#   - This file doesn't have a header.

#   - Use flag: --pf


# ### Input parameters:
# 1. Training sample size. 
#   - Use flag: --N
# 2. Estimated SNP Heritability (pre-compute this using your favorite method). 
#   - Use flag: --H2
# 3. LD radius (optional). If not provided, it is computed as (1/2)*0.15% of total number of SNPs.  
#   - Use flag: --ld_radius

# ### Output files
# 1. Coordinated files: This is an hdf5 file that stores the coordinated genotype data with the summary statistics and functional enrichments.
#   - Use flag: --coord
#   - Note: the output file needs to be named differently for different runs.
  
# 2. Posterior mean effect sizes: Estimated posterior mean effect size from LDpred-funct-inf.
#   - Use flag: --posterior_means
  
# 3. Output: Polygenic risk score for each individual in the validation. 
#   - Description:
#     - Column 1: Sample ID
#     - Column 2: True phenotype
#     - Column 3: PRS using all-snps and marginal effect sizes.
#     - Colunm 4: PRS obtained using LD-pred-funct-inf
#     - Column 5-K: PRS(k) defined in equation 5 from Marquez-Luna, et al, Biorxiv.

#   - Use flag: --out
  
### Example
```

plinkfile="../private_input_files/BB_imputed_data_sample/ped/BB_imputed_data_sample[1:22]"
outCoord="my_analysis/Coord_Final"
statsfile="my_analysis/sumary_statistics.txt"
functfile="my_analysis/functfile_sldsc.txt"
outLdpredfunct="my_analysis/ldpredfunct_posterior_means"
outValidate="my_analysis/ldpredfunct_prs"
phenotype="my_analysis/trait_1.txt"
N=training sample size
h2= pre-computed SNP heritability

# optional flags:
# --ld_radius= Pre-defined ld-radius
# --K= Number of bins for LDpred-funct

python LDpred-funct/ldpredfunct.py --gf=${plinkfile} --pf=${phenotype} --FUNCT_FILE=${functfile}  --coord=${outCoord} --ssf=${statsfile} --N=${N} --posterior_means=${outLdpredfunct}  --H2=${h2} --out=${outValidate} > ${outValidate}.log
```